src/main/java/data/TaskList.java

```diff
@@ -0,0 +1,153 @@
+     * @throws InvalidTaskException if i < 0  or i > taskList.getSize()
+     */
+    public void deleteTask(int i, Ui ui) throws InvalidTaskException {
+        int numberOfTasks = this.taskList.size();
+        if (i <= 0 || i > numberOfTasks) {
+            throw new InvalidTaskException("Task does not exist!");
+        }
+        Task task = this.taskList.get(i - 1);
+        this.taskList.remove(i - 1);
+        numberOfTasks--;
+
+        ui.displayString("Deleted: " + task);
+        ui.displayString("You now have " + numberOfTasks + " tasks.");
+    }
+
+    /**
+     * Marks task as complete. Shows user success message upon completion.
+     *
+     * @param i task number of the task to be marked
+     * @param ui user interface to give user information
+     * @throws InvalidTaskException if i < 0  or i > taskList.getSize()
+     */
+    public void markTask(int i, Ui ui) throws InvalidTaskException {
+        int numberOfTasks = this.taskList.size();
+        if (i <= 0 || i > numberOfTasks) {
+            throw new InvalidTaskException("Task does not exist!");
+        }
+        Task task = this.taskList.get(i - 1);
+        task.markDone();
+
+        ui.displayString("Good Job! The task is now marked as done: ");
+        ui.displayString("Marked task: " + task);
+    }
+
+    /**
+     * Marks task as not complete. Shows user success message upon completion.
+     *
+     * @param i task number of task to be unmarked
+     * @param ui user interface to give user information
+     * @throws InvalidTaskException if i < 0  or i > taskList.getSize()
+     */
+    public void unmarkTask(int i, Ui ui) throws InvalidTaskException {
+        int numberOfTasks = this.taskList.size();
+        if (i <= 0 || i > numberOfTasks) {
+            throw new InvalidTaskException("Task does not exist!");
+        }
+        Task task = this.taskList.get(i - 1);
+        task.markNotDone();
+
+        ui.displayString("Alright, the task is marked as not done: ");
+        ui.displayString("Unmarked task: " + task);
+    }
+
+    /**
+     * Lists all tasks and shows them to user.
+     *
+     * @param ui user interface to show user information
+     */
+    public void listAllTasks(Ui ui) {
+        int numberOfTasks = taskList.size();
+        ui.displayString("You currently have " + numberOfTasks + " tasks.");
+        for (int i = 0; i < numberOfTasks; i++) {
+            int num = i + 1;
+            Task task = taskList.get(i);
+            ui.displayString(num + ". " + task);
+        }
+    }
+
+    /**
+     * Method to search and display tasks based on keyword given
+     *
+     * @param searchStr keyword to search tasks with
+     * @param ui user interface to give user information
+     */
+    public void searchTasks(String searchStr, Ui ui) {
+        int numberOfTasks = taskList.size();
+        int num = 1;
+        ui.displayString("Here are the tasks with keyword '" + searchStr +"':");
```

Comment:
```suggestion
        ui.displayString("Here are the tasks with keyword '" + searchStr + "':");
```

I think operators must be followed by whitespace.